README.md
## npm script

``` shell
#启动服务
npm run serve

#打包编译 
npm run build

#执行所有测试用例
npm run test 

#执行单个测试用例(输入测试用例文件名即可，无需输入后缀)
npm run test ${yourfileName} 
```

CHANGELOG.md
## [1.0.67-alpha.5](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.67-alpha.4...v1.0.67-alpha.5) (2021-06-04)



## [1.0.67-alpha.4](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.67-alpha.3...v1.0.67-alpha.4) (2021-05-26)



## [1.0.67-alpha.3](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.67-alpha.2...v1.0.67-alpha.3) (2021-05-26)



## [1.0.67-alpha.2](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.67-alpha.1...v1.0.67-alpha.2) (2021-05-26)



## [1.0.67-alpha.1](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.67-alpha.0...v1.0.67-alpha.1) (2021-05-26)



## [1.0.67-alpha.0](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.66...v1.0.67-alpha.0) (2021-05-25)



## [1.0.66](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.66-alpha.9...v1.0.66) (2021-05-25)



## [1.0.66-alpha.9](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.66-alpha.8...v1.0.66-alpha.9) (2021-05-25)



## [1.0.66-alpha.8](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.66-alpha.7...v1.0.66-alpha.8) (2021-05-25)



## [1.0.66-alpha.7](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.66-alpha.6...v1.0.66-alpha.7) (2021-05-25)



## [1.0.65](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.64...v1.0.65) (2021-05-14)


### Bug Fixes

* **date-helper:** 修复getTimeInterval方法，秒级时取整 ([2dc68aa](http://git.nextop.cc/FE-Lab/vantop-util/commits/2dc68aac4deaf497ecb91c234147740d9f5f4d3d))



## [1.0.64](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.64-alpha.0...v1.0.64) (2021-05-11)



## [1.0.64-alpha.0](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.63...v1.0.64-alpha.0) (2021-05-11)



## [1.0.63](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.62...v1.0.63) (2021-05-07)



## [1.0.62](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.62-alpha.0...v1.0.62) (2021-05-07)



## [1.0.62-alpha.0](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.61...v1.0.62-alpha.0) (2021-05-07)



## [1.0.61](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.61-alpha.3...v1.0.61) (2021-04-19)


### Features

* **general:** 新增按需引入功能，并删除Helper、DateHelper、MathHelper、Validator的用法 ([cbee4f8](http://git.nextop.cc/FE-Lab/vantop-util/commits/cbee4f824457853f94b5f7148e5f1d229bac55a0))



## [1.0.61-alpha.3](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.61-alpha.2...v1.0.61-alpha.3) (2021-04-19)



## [1.0.61-alpha.2](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.61-alpha.1...v1.0.61-alpha.2) (2021-04-16)



## [1.0.61-alpha.1](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.61-alpha.0...v1.0.61-alpha.1) (2021-04-12)



## [1.0.61-alpha.0](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.60...v1.0.61-alpha.0) (2021-04-12)



## [1.0.60](http://git.nextop.cc/FE-Lab/vantop-util/compare/v1.0.59...v1.0.60) (2021-03-18)


### Bug Fixes

* **browser:** 拷贝函数正则表达式在Safari浏览器异常，已修复拷贝函数方式 ([b59b454](http://git.nextop.cc/FE-Lab/vantop-util/commits/b59b4546daac8b0eee01f9c3eee7f3b06960cace))



## [1.0.59](https://github.com/sharly1201/vantop/compare/v1.0.58...v1.0.59) (2021-03-05)


### Features

* **helper:** 新增hasOwnProperty方法 ([13ce262](https://github.com/sharly1201/vantop/commit/13ce26207eceb7cb264d6784c41fbafd4c92ffc1))


### Performance Improvements

* **math-helper:** 数学助手代码优化 ([7fb303c](https://github.com/sharly1201/vantop/commit/7fb303c1a39da4b75b72650003d443f6c2f8c6b6))



## [1.0.58](https://github.com/sharly1201/vantop/compare/v1.0.57...v1.0.58) (2021-03-03)



## [1.0.57](https://github.com/sharly1201/vantop/compare/v1.0.57-alpha.5...v1.0.57) (2021-03-03)


### Features

* **helper:** add helper/removeInvisibleCharacter ([6fcc33a](https://github.com/sharly1201/vantop/commit/6fcc33a4631ea16c0df9a275656d2ceee91a94cf))
* **vantop-util:** 添加localStorage和sessionStorage浏览器存储方法 ([eb4ff0d](https://github.com/sharly1201/vantop/commit/eb4ff0de92862b264c674e1ed86e6e9ad574f55e))



## [1.0.57-alpha.5](https://github.com/sharly1201/vantop/compare/v1.0.57-alpha.4...v1.0.57-alpha.5) (2021-02-23)


### Features

* **vantop-util:** test changelog ([236d614](https://github.com/sharly1201/vantop/commit/236d614a6dadc11c0706d333df3ff92103371ebb))



## [1.0.57-alpha.2](https://github.com/sharly1201/vantop/compare/v1.0.57-alpha.1...v1.0.57-alpha.2) (2021-02-08)


### Bug Fixes

* **vantop-util:** test changelog ([687eae7](https://github.com/sharly1201/vantop/commit/687eae735ef277b69771fd3f02e355ae1f31f78c))


### Features

* **vantop-util:** test changelog ([ed87763](https://github.com/sharly1201/vantop/commit/ed87763e4b4059d5b7bf0870ae04b2e4d98a85d1))



## [1.0.57-alpha.1](https://github.com/sharly1201/vantop/compare/v1.0.57-alpha.0...v1.0.57-alpha.1) (2021-02-08)



## 1.0.57-alpha.0 (2021-02-08)



# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [1.0.56](https://github.com/sharly1201/vantop/compare/@vantop/vantop-util@1.0.55...@vantop/vantop-util@1.0.56) (2021-02-05)

**Note:** Version bump only for package @vantop/vantop-util





## [1.0.55](https://github.com/sharly1201/vantop/compare/@vantop/vantop-util@1.0.54...@vantop/vantop-util@1.0.55) (2021-02-03)

**Note:** Version bump only for package @vantop/vantop-util





## [1.0.54](https://github.com/sharly1201/vantop/compare/@vantop/vantop-util@1.0.52...@vantop/vantop-util@1.0.54) (2021-02-03)


### Features

* **vantop-util:** th新增throttle,修复debounce的this指向问题，cookie方法名修改 ([17b8c0b](https://github.com/sharly1201/vantop/commit/17b8c0bfb362f408dd990ca3d299fab4bf64472d))


### Reverts

* **documentations:** test ([7d0951a](https://github.com/sharly1201/vantop/commit/7d0951a3b1684775cd729630ea623130f4bb0963))
* **vantop-ui:** clear package-lock.json ([3fa4471](https://github.com/sharly1201/vantop/commit/3fa44713048a88396e47ed6e25f0b6fffa7719f5))
* **vantop-util:** clear is number ([0047007](https://github.com/sharly1201/vantop/commit/00470079c287e0b4ec4fd0d56ac357d052a2fb6d))






src/browser/userAgent.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/*
 * @Author: 易鑫
 * @Date: 2021-01-25 18:10:03
 * @LastEditTime: 2021-04-14 19:09:55
 * @Description: 浏览器UA检测
 */
const getUserAgent = (ua?: string): string => {
    if (typeof window !== 'object') {
        if (ua) {
            return ua;
        }
        throw new ReferenceError('[Reference Error] Please run UA in a browser environment!');
    }
    return window.navigator.userAgent;
};

const isFirefox = (ua?: string): boolean => getUserAgent(ua).indexOf('Firefox') > -1;

const isWebKit = (ua?: string): boolean => getUserAgent(ua).indexOf('WebKit') > -1;

const isChrome = (ua?: string): boolean => getUserAgent(ua).indexOf('Chrome') > -1;

const isOpera = (ua?: string): boolean => getUserAgent(ua).indexOf('Opera') > -1;

const isIE = (ua?: string): boolean => getUserAgent(ua).indexOf('compatible') > -1 && getUserAgent(ua).indexOf('MSIE') > -1 && !isOpera(ua);
//IE内核
const trident = (ua?: string): boolean => getUserAgent(ua).indexOf('Trident') > -1;
//opera内核
const Presto = (ua?: string): boolean => getUserAgent(ua).indexOf('Presto') > -1;
//苹果、谷歌内核
const webKit = (ua?: string): boolean => getUserAgent(ua).indexOf('AppleWebKit') > -1;
//火狐内核gecko
const gecko = (ua?: string): boolean => getUserAgent(ua).indexOf('gecko') > -1;
//是否为移动终端
const mobile = (ua?: string): boolean => !!getUserAgent(ua).match(/AppleWebKit.*Mobile.*/);

const isIOS = (ua?: string): boolean => !!getUserAgent(ua).match(/\(i[^;]+;( U;)? CPU.+Mac OS X/);

const isAndroid = (ua?: string): boolean => getUserAgent(ua).indexOf('Android') > -1 && getUserAgent(ua).indexOf('Linux') > -1;

const iPhone = (ua?: string): boolean => getUserAgent(ua).indexOf('iPhone') > -1;

const iPad = (ua?: string): boolean => getUserAgent(ua).indexOf('iPad') > -1;

const isWechat = (ua?: string): boolean => getUserAgent(ua).indexOf('MicroMessenger') > -1;

const isMobile = (ua?: string): boolean => mobile(ua) || isIOS(ua) || isAndroid(ua) || iPhone(ua) || iPad(ua);

const isPc = (ua?: string): boolean => {
    const userAgentInfo = getUserAgent(ua);
    const agents = ['Android', 'iPhone', 'SymbianOS', 'Windows Phone', 'iPad', 'iPod'];
    let flag = true;
    for (let v = 0; v < agents.length; v++) {
        if (userAgentInfo.indexOf(agents[v]) > 0) { flag = false; break; }
    }
    return flag;
};

const UA = Object.freeze({
    isPc,
    isFirefox,
    isChrome,
    isOpera,
    isIE,
    isMobile,
    isIOS,
    isAndroid,
    isWechat,
});

export default UA;